Build Stdlib with Noncopyable Generics (Part 4)

kavon opened the pull request:
This PR makes it so you can build the compiler and stdlib with `--enable-experimental-noncopyable-generics=1` given to build-script. This allows you to properly test and experiment with noncopyable generics using the stdlib types.

kavon committed:
[NCGenerics] trigger module mismatch

A swiftmodule can only be correctly ingested by a compiler
that has a matching state of using or not-using
NoncopyableGenerics.

The reason for this is fundamental: the absence of a Copyable
conformance in the swiftmodule indicates that a type is
noncopyable. Thus, if a compiler with NoncopyableGenerics
reads a swiftmodule that was not compiled with that feature,
it will think every type in that module is noncopyable.

Similarly, if a compiler with NoncopyableGenerics produces a
swiftmodule, there will be Copyable requirements on each
generic parameter that the compiler without the feature will
become confused about.

The solution here is to trigger a module mismatch, so that
the compiler re-generates the swiftmodule file using the
swiftinterface, which has been kept compatible with the compiler
regardless of whether the feature is enabled.

kavon committed:
[NCGenerics] fix hack in `requiresInvertible`

kavon committed:
[build-script] fix NoncopyableGenerics option

This patch introduces `--enable-experimental-noncopyable-generics` for
the build script. It replaces
`--swift-stdlib-experimental-noncopyable-generics`

The old build option only enables the feature when building the
stdlib, but if we've built the stdlib with NoncopyableGenerics, the
compiler should be hardwired to have that feature enabled, too.

This patch also introduces the `noncopyable_generics` lit parameter, so
that tests assuming the system was built with the feature can live
in-tree and be tested, if they specify `REQUIRES: noncopyable_generics`.

kavon commented:
@swift-ci please test

kavon committed:
[Mangling] do not mangle invertible protocols

Requirements for invertible protocols are always assumed to exist
unless opted-out via an inverse requirement. Thus, it's the
absence of those requirements that will (eventually) get mangled
into symbols.

kavon commented:
@swift-ci please test

kavon committed:
[SILGen] tweak LogicalKeyPathApplicationComponent

A baked-in assumption was that the generic parameters of the KeyPath
family of types have no conformance requirements. That's no longer the
case with NoncopyableGenerics, so we need to dig up conformances when
forming the substitution map, as-needed.

kavon committed:
[IRGen] fix `createCXXCopyConstructorFunctionType`

The previous implementation assumes `UnsafePointer` has no requirements
on its generic parameter. This fix handles any requirements that may be
present on the parameter. With NoncopyableGenerics, the generic
parameter actually does gain a requirement.

kavon committed:
[NCGenerics] enable new associated type inference

The experimental-associated-type-inference flag is needed to correctly
build programs when NoncopyableGenerics is enabled.

kavon committed:
[NCGenerics] SwiftSources == CompilerInvocation

Match up the additional flags in SwiftSources.cmake and
CompilerInvocation.cpp with respect to NoncopyableGenerics.

I think only CompilerInvocation is truly needed, though.

kavon committed:
[NFC] replace `canBeNoncopyable`

First, "can have an absence of Copyable" is a rather confusing notion,
so the query is flipped to "can be Copyable". Next, it's more robust to
ask if a conformance exists for the TypeDecl to answer that question,
rather than trying to replicate what happens within that conformance
lookup.

Also renames `TypeDecl::isEscapable` to match.

kavon committed:
[NCGenerics] remove arg from `isNoncopyable(DC)`

It should be the responsibility of callers to map the type to a
contextual type, as needed. When it's not possible or repetitive to do
so, there is a special-purpose function `isInterfaceTypeNoncopyable` for
 Sema.

kavon committed:
[NCGenerics] hack for module mismatch

kavon committed:
[NCGenerics] require associated type inference when rebuilding module

slavapestov commented on lib/SILGen/SILGenLValue.cpp:
```diff
@@ -2330,21 +2331,31 @@ namespace {
+                                                  ->castTo<SubstitutableType>();
+      map[firstGenParam] = BaseFormalType;
+
+      auto subs = SubstitutionMap::get(sig,
```
It might be easier to use the get() that just takes arrays of replacement types and conformances

kavon replied:
In terms of lines-of-code, this version building a TypeSubstitutionMap ended up being more concise / clean.

slavapestov commented on lib/IRGen/GenStruct.cpp:
```diff
@@ -597,9 +597,18 @@ namespace {
+      TypeSubstitutionMap map;
+      map[param] = T.getASTType();
+
+      auto subst = SubstitutionMap::get(sig,
```
Same remark about SubstitutionMap::get()

slavapestov commented on include/swift/AST/Decl.h:
```diff
@@ -3175,13 +3175,33 @@ class TypeDecl : public ValueDecl {
-  /// Is it possible for this type to lack a Copyable constraint?
+  /// Indicates how "strongly" a TypeDecl will conform to an invertible
+  /// protocol. Supports inequality comparisons and casts to bool.
+  enum CanBeInvertibleResult: unsigned {
```
Use an enum class so that the case names are namespaced properly

kavon replied:
I wrapped the enum in a class for the namespacing, while preserving the behavior that the result of `canBeCopyable` can be cast to bool implicitly. Might be a little gross but it does make the code pretty. Since the conditional case isn't yet being used as I thought, perhaps it should be renamed (again) to `alwaysCopyable` since the conditionally copyable case is always lumped into the not-copyable situations, especially in type lowering.

kavon commented:
@swift-ci smoke test

kavon commented:
@swift-ci please smoke test

kavon commented:
https://github.com/apple/llvm-project/pull/7992
@swift-ci please test

kavon committed:
[NCGenerics] provide namespace for CanBeInvertible

kavon commented:
https://github.com/apple/llvm-project/pull/7992
@swift-ci please test

kavon committed:
[NCGenerics] fix Combine issue

The solution is to add the unconditional conformances when building the
conformance lookup table for each nominal, instead of waiting until
conformance lookup infers it.

At the moment, because the stdlib isn't being built with Copyable
conformances injected into it, we still end up needing to infer Copyable
conformances for nominals like `Int` from the stdlib when trying to test
the feature, since it loads those nominals, but they lack conformance.

That should go away once we do enable NoncopyableGenerics for the stdlib
by default, and we'll be left with only inferring the conditional
conformance in TypeCheckInvertible.

kavon committed:
[NCGenerics] Sendable doesn't require Copyable

kavon committed:
[test] coverage for conditional conformances

kavon committed:
[NCGenerics] fix a few `isNoncopyable` calls

These calls were being made on types either with a type parameter, or a
SIL-only type that doesn't actually have conformances.

In such cases, it's better to use SILType's move-only query methods.

kavon committed:
SILType: Avoid asking ASTType if it is Noncopyable

It's better to ask SILType if it is MoveOnly than go to the AST type and
 ask if it is noncopyable, because some types in SIL do not have a
 well-defined notion of conformance in the AST.

kavon committed:
NFC: describe assertion issue

kavon committed:
Sema: have createBuiltinCall handle conformances

The only callers of `DerivedConformance::createBuiltinCall` aren't
actually aware of what the generic signature is for the builtin they
want to call. Thus, they can't really provide the needed conformances.

This wasn't an issue when there were no conformances needed for the
substitution map, but with NoncopyableGenerics, we need at least
Copyable conformances for the generic parameters of a builtin. It's easy
enough to make this function just handle generating the conformances
list, so that's the refactoring here.

slavapestov commented on lib/Sema/DerivedConformances.cpp:
```diff
@@ -472,8 +470,19 @@ DerivedConformance::createBuiltinCall(ASTContext &ctx,
+      map.insert({genericParams[i]->getAs<SubstitutableType>(), typeArgs[i]});
+
+    auto subs = SubstitutionMap::get(generics,
+                                     QueryTypeSubstitutionMap{map},
```
A bunch of usages of QueryTypeSubstitutionMap could instead use an array of generic arguments in the correct order instead. If you want, you can add a new functor, like QueryReplacementTypeArray or something that does this. To map between from a generic parameter type to array index, you can use GenericParamKey:

```
  GenericParamKey key(paramTy);
  unsigned index = key.findIndexIn(genericParams);
```

kavon replied:
I tried this briefly but the functor only receives a SubstitutableType and it wasn't obvious how to derive the `GenericParamKey` for one of those. For now I'll leave this in as using `QueryTypeSubstitutionMap` but I agree with what you're getting at here... `QueryTypeSubstitutionMap` is a bit unwieldy and if you don't build it with a canonical type of the generic parameter as the key, you silently get no substitutions.

kavon replied:
OK I can do this clean-up in a follow-up PR along with simplifying CollectExistentialConformances.

kavon committed:
NCGenerics: simplify `isInterfaceTypeNoncopyable`

We can greatly simplify our noncopyable checks if we skip running these
diagnostic checks while in SIL mode, where it's not really needed. The
challenge with doing the checks in SIL mode is that there is sometimes
no generic environment when doing these queries. Whether that itself is
simply a bug or fundamental limitation, I haven't investigated.

kavon commented:
https://github.com/apple/llvm-project/pull/7992
@swift-ci please smoke test

kavon commented:
https://github.com/apple/llvm-project/pull/7992
@swift-ci please smoke test

kavon commented:
https://github.com/apple/llvm-project/pull/7992
@swift-ci please smoke test

kavon committed:
NCGenerics: adjust number of diagnostics

Something I changed in the implementation of NoncopyableGenerics has
gotten the number back down to 6, where it was before I started :)

kavon commented:
https://github.com/apple/llvm-project/pull/7992
@swift-ci please test

kavon commented:
https://github.com/apple/llvm-project/pull/7992
@swift-ci Please Test Source Compatibility

kavon committed:
NCGenerics: isMoveOnly and reference ownership

ReferenceStorageType can only contain a type that is class-like, like a
SILBox or a class-constrained existential, etc.

While classes do not and are not planned to officially support being
noncopyable, we do still have a lingering experiemental feature that
allows classes to be marked `@_moveOnly`. Thus, the best way to handle
queries about a `ReferenceStorageType` being move-only or noncopyable is
 to check whether the referent is.

kavon committed:
NCGenerics: adjust for mangling & type lowering

We're no longer mangling the Copyable or Escapable protocol into generic
signatures, but haven't yet started mangling inverses into it, thus one
function got renamed

Otherwise, the type lowering has become more pessimistic about whether
the type is actually Copyable, thus a few functions have changed their
convention to be passed indirectly (`@guaranteed`).

kavon commented:
https://github.com/apple/llvm-project/pull/7992
@swift-ci please test

kavon commented:
https://github.com/apple/llvm-project/pull/7992
@swift-ci please test

kavon commented:
https://github.com/apple/llvm-project/pull/7992
@swift-ci please test

slavapestov commented on lib/Sema/TypeCheckType.cpp:
```diff
@@ -2371,7 +2371,7 @@ bool swift::diagnoseMissingOwnership(ASTContext &ctx, DeclContext *dc,
     return false; // no need for ownership in enum cases.
 
-  if (!ty->isNoncopyable(dc))
+  if (!isInterfaceTypeNoncopyable(ty, dc->getGenericEnvironmentOfContext()))
```
use options.getGenericSignature().getGenericEnvironment() instead -- that will fix SIL mode.

slavapestov commented on lib/Sema/TypeCheckType.cpp:
```diff
@@ -4776,7 +4777,9 @@ NeverNullType TypeResolver::resolveVarargType(VarargTypeRepr *repr,
-      && element->isNoncopyable(getDeclContext())) {
+      && !options.contains(TypeResolutionFlags::SILMode)
+      && isInterfaceTypeNoncopyable(
+          element, getDeclContext()->getGenericEnvironmentOfContext())) {
```
This should also be changed to options.getGenericSignature().getGenericEnvironment()

kavon commented:
tests on macOS all passed, just an unrelated test that is failing in CI generally at the time of testing

kavon commented:
https://github.com/apple/llvm-project/pull/7992
@swift-ci please smoke test macOS

kavon commented:
https://github.com/apple/llvm-project/pull/7992
@swift-ci please smoke test macOS

kavon commented:
https://github.com/apple/llvm-project/pull/7992
@swift-ci please smoke test macOS

kavon commented:
https://github.com/apple/llvm-project/pull/7992
@swift-ci please smoke test macOS
